### DISCUSSION TITLE
APIs for playlist manipulation would be nice

### DISCUSSION THREAD

**Discussion by eras** [tidal-music/discussions#6] [Ideas & Feature requests]
This would allow buildings tools like playlist organizers. In my opinion organizing playlists is a bit lacking in the current TIDAL interface.

**Comment by aliaksei-taliuk** [comment]
Hi @eras , welcome to the forum, and thanks for your suggestion!

Exposing API for playlist manipulation is definitely on our TODO list. We can unfortunately not promise any date for when this will be available, but we will continue to share any progress we make.

Thanks

**Reply by schnetzi** [reply]
Are you able to provide more insights on your project plans? So what the next features will be and at what priority this feature is?

For me this is also a very important feature as I want to sync my playlists from spotify to tidal via node.js and I haven't found the possibility to do so with tidal.

**Reply by stefanmoro** [reply]
Hi @schnetzi,

TIDAL Developer Platform is still in very early stages, and we have taken an approach where we make things available as soon as we have something new to share, rather than develop it closed and release once "feature complete".

We hope that this approach will lead to faster feedback from the community, and ultimately better solutions. But we also acknowledge that it can lead to us not being able to give a clear timeline for all requests. 

However, accessing "user data" (playlists, favourites etc) is one feature we are actively looking in to. Once we have any concrete updates we will make sure to share them here.

**Comment by einaralex** [comment]
Definitely, with **collaborative playlists** in mind, please!

**Reply by lyzkov** [reply]
Sure!

**Comment by thephenakist** [comment]
An ideal playlist endpoint would provide (at least) the ability to:
- Get Playlist (GET, information about the playlist, description, image, created date, owner...)
- Update Playlist (PUT, image, description...)
- Create Playlist (PUT, if no playlist ID specified a new playlist is created and the new playlist ID is returned in response)
- Get Playlist Items (GET, in addition to details provided by Get Playlist, this would also return the track IDs for items within the playlist)
- Add Items To Playlist (PUT, at a specified 'index' if specified, else index = length of items currently in playlist)
- Remove Items From Playlist (PUT, either by index w/ length of items, or by supplying IDs to remove from the playlist)
- Replace Playlist Items (PUT, all current items will be replaced by the supplied list of items)
- Get Playlist Version (GET, It may be beneficial to keep a unique version ID for the playlist, which clients can cache on calls and reference later to determine if they are in-sync with TIDAL's servers. This could simply be provided in the Get Playlist call, but worth mentioning.)
- Get User Playlists (GET, returns all playlists where specified user id is the creator / owner)

**Reply by fmalhe** [reply]
Hi @thephenakist, Thanks for the suggestion!
We are planning to get started on this soon, and community feedback is very welcome. Will post here once we have some more updates to share.

**Reply by gigbuilder** [reply]
Can you define "soon"

**Comment by ngoreico** [comment]
Hi ! 
I was wondering if there is any endpoint available right now to add songs to a playlist, considering that Tidal suggests using a third-party application to add songs from another platform to your account.
I would like to do the same thing as TuneMyMusic by my own.
Thanks !

**Reply by stefanmoro** [reply]
Hey @ngoreico 

There are currently no endpoints available for this, but we are hoping to start this work soon!

**Comment by jadewilk** [comment]
Hello, I'm making a small program in rust for moving user's playlists from spotify to other streaming services using the ISRC of each track as the common ID across all services - so far I've implemented spotify to youtube, but I want to add Tidal as another destination. For this, I need to be able to get tracks by their ISRC, which you already have implemented, but also need to be able to create playlists and add tracks to them.

I have a suggestion for whatever endpoint you decide to have for adding tracks to a playlist; please support adding multiple tracks per api call. While this isn't much of a problem for me developing against the api; I can just have a for loop that iterates over each track and calls the endpoint for each one individually, it's admittedly very wasteful when it could just be batch processed with minimal requests. This is important for devs if you charge for using the api per request.

Youtube does not support this so the number of requests ends up being very high, which is especially important in their case where you have a quota and each request has a cost. [Spotify currently supports this ](https://developer.spotify.com/documentation/web-api/reference/add-tracks-to-playlist)kind of action up to 100 tracks at a time, which is a fairly reasonable limit; a playlist of say 2500 tracks on takes 25 requests rather than 2500.

Also, would the saved tracks in 'my collection' be treated as a playlist or would they have their own endpoint?

**Reply by stefanmoro** [reply]
Hi @philipwilk ,

Our current priority is to expose more and unify existing data domains (my collection, user, playlists, revamped catalogue) on TIDAL Developer, so very timely input, thank you!

We will take your input into consideration as we now build out this access, and we hope to open up this functionality within the coming months.

**Reply by Kubas556** [reply]
Theoretically it is possible to use unofficial endpoint "https://listen.tidal.com/v2/my-collection/playlists/folders/create-playlist".

**Comment by dylmye** [comment]
I'm excited for this - I've been building [a concert setlist app](https://github.com/dylmye/setlist-sherlock) and my use case for this API is adding the tracks from a setlist to a user's playlist. I've been able to do this with Spotify's API but they aren't approving anyone's production usage this year 😬 Tidal is the superior platform so 🤞 this is on their roadmap soon!

**Reply by lyzkov** [reply]
Spotify has a terrible mobile SDK. They cut off most of the valuable partnerships couple of years ago with one unconsulted decision about discontinuing streaming services to integrated players. TIDAL has the potential to fill an abandoned niche.

**Comment by trbryant** [comment]
Also interested in this would like to build a workflow that accepts a request (artist name, track name) and then performs a search which returns a result set which can then be added to a playlist.

**Comment by peavers** [comment]
Simply being able to read a users collections/playlists would be enough for me at this point in time. I have user oauth flows wired up, but nothing to invoke with them :(

**Comment by gigbuilder** [comment]
Its been a year and we still do not have any method to generate or populate playlists.  I write a CRM for DJs and we have about 2,000 customers - many want to use Tidal but the API is so limited that we really can't do anything to help them.  When are you going to provide something similar to Spotify?  They did it right.

**Comment by stefanmoro** [comment]
Hey,

A couple of months back we started rolling out our "V2" endpoints, essentially an improved way to format and structure our endpoints - mainly in how we handle relationships. We started with [catalogue](https://developer.tidal.com/apiref?spec=catalogue-v2) and [search](https://developer.tidal.com/apiref?spec=search-v2), and over the last 2 months we have also started exposing [user](https://developer.tidal.com/apiref?spec=user-v2) data.

Just last week we rolled out the first endpoints giving access to [playlists](https://developer.tidal.com/apiref?spec=user-playlist-v2). At this stage we allow read access to user playlists, but we will start work on building support for write operations relatively soon.

The entire API is still in "beta", and will remain so for some more time as we continue to build out our new API, and also to integrate it into our own apps.

In addition to write operations for playlists, we are also planning for read/write operations to user collections.

**Reply by dylmye** [reply]
This is great news! Thank you to everyone involved in this work!

**Reply by thephenakist** [reply]
This is not currently working for me, neither in my project nor using API Explorer Mode. Using the latter produces a 400 error with the response `Required parameter 'userId' is not present.` (not possible to enter a value using the web interface), and in my code if I pass a userId param it seems to generate a 500 server error. Let me know if I can be of assistance debugging the issue, but I presume this is a known issue at the moment for playlists.

**Reply by aliaksei-taliuk** [reply]
@thephenakist Thanks for reaching out. Could you please provide more context about your problem (request details, auth method, client id, etc)? Thanks.

**Reply by thephenakist** [reply]
@aliaksei-taliuk As for API Explorer Mode, I'm using [Get single playlist](https://developer.tidal.com/apiref?spec=user-playlist-v2&ref=get-playlist-by-id-v2) (i.e. click "Try It Out", authorize with client ID and secret and populate the required params of Country Code and ID). I get the following response:
```
body:
{
  "type": "about:blank",
  "title": "Bad Request",
  "status": 400,
  "detail": "Required parameter 'userId' is not present.",
  "instance": "/playlists/7181f3e0-30ed-4dce-9154-f1ab493d89d6"
}

headers:
 cache-control: max-age=0,proxy-revalidate 
 content-length: 173 
 content-type: application/problem+json 
```
In my code I'm generating a token using clientId: 'aq7KxY3lSag82aOh' (and secret) against `auth.tidal.com` and then sending the token along with the authorization header for `Bearer`. My request uses the following details along with the bearer token:
```
url: 'openapi.tidal.com/v2/playlists/7181f3e0-30ed-4dce-9154-f1ab493d89d6'
countryCode: 'US'
include: 'items'
```
I've tried with and without `include: 'items'`, with and without `userId: 195828142`, for both public and private playlists. If `userId` is not present I get the same response as API Explorer Mode above. With `userId` present I get a 500 error.

**Reply by stefanmoro** [reply]
Hey @thephenakist ,

Since a playlist is owned by a user, the owner has to authorize any app wanting to access the playlist access to it. This is done using OAuth authorization code flow and scopes. You are trying to access playlists using an access token acquired through a client credentials flow, which will not work.

TIDAL Developer Platform uses OAuth 2.1 to handle authorization. Please read our [authorization documentation](https://developer.tidal.com/documentation/authorization/authorization-overview) for further information on how to correctly configure your app, and grant it access to playlists using authorization code flow.

Thanks

**Reply by thephenakist** [reply]
Ah, thank you. I understand the design behind this endpoint now (that is: for users who have authorized an application permission to read the contents of THEIR OWN playlists). In the future I would like the ability for my application to READ the contents (items, description, owner, etc) of PUBLIC playlists, regardless of creator or app authorizations by using the playlist UUID.

**Reply by stefanmoro** [reply]
Hey @thephenakist,

Thank you for explaining your use case further, I get what you mean now. 

You are correct, the current playlist endpoints require a request to be made on behalf of a TIDAL user (i.e. they require the supplied access token to be acquired through authorization code flow), regardless of if the playlist is public or not. 

We have briefly discussed the possibility to allow "client credentials" access to public playlists, and will pick up that discussion as we continue expanding our playlist endpoints later this year. Good to hear you have a use case for it, valuable input for us!

**Reply by 1729patrick** [reply]
Hi @stefanmoro,

I'm currently implementing your API in my app and am excited about the upcoming playlist creation feature. However, I'm encountering an issue with retrieving the user's playlists. 

In the API Explorer Mode, I can see all my playlists. Yet, when I run the authorization flow to obtain user permissions (`playlists.write` and `playlists.read`), and use that token to fetch the playlists, the response returns an empty array of items. 

I’ve debugged the access token and noticed one difference: the presence of `"cuk": "developerPortal"` and `kid`.

API Explorer Mode
<img width="581" alt="image" src="https://github.com/user-attachments/assets/271ac0e7-e22b-4acd-a0e7-b4e2e0251a99"> 

User Authorization Flow
<img width="578" alt="image" src="https://github.com/user-attachments/assets/4afe8f10-e3d7-4e45-964e-af346df60524">

Could you please help me understand why this might be happening?

Thank you!

**Reply by 1729patrick** [reply]
Just a follow-up on the previous issue @stefanmoro. It turns out the problem was related to the scopes. Having multiple scopes, even including `"playlists.read"`, was causing the API to return an empty list of playlists for the user. 

I resolved the issue by requesting only the permission to read users' playlists.

**Reply by aliaksei-taliuk** [reply]
@1729patrick This is also fixed already. Should be working for multiple scopes.

**Reply by januszmk** [reply]
@stefanmoro is there any update on when are you gonna rollout API endpoint for write operations to user playlists? I would love to be able to generate my own playlists automatically...

**Comment by jasongrishkoff** [comment]
Hi there, just chiming in with my use-case for the ability to manipulate playlists. We're implementing "playlist syncing" on SubmitHub.com, allowing the users to automatically keep their playlists synced cross-platform. So far we've integrated Apple, YouTube, Deezer, and SoundCloud. I've had a few curators ask me to add Tidal support, so will be keeping an eye on this thread :)

**Comment by smoothquark** [comment]
Thank you for this. I have been playing with the API. Quick question please - are the number of items returned limited to a maximum of 20? Am I missing something?
Examples: running from https://developer.tidal.com/apiref?spec=catalogue-v2&ref=get-artist-albums-v2 (https://openapi.tidal.com/v2/artists/{artist}/relationships/albums?countryCode=GB&include=albums)
1. id 1566 (Beyoncé) generates 20 albums on the API and also 20 on listen.tidal.com -> albums -> view all
2. Herbert von Karajan (64046) the api generates 20 albums as well but on listen.tidal.com -> albums -> view all there are well over 200 listed
I am pretty sure Beyoncé has more than 20 albums on Tidal but only 20 are listed with both the API and listen.tidal.com, but Herbert von Karajan has 20 vs > 200?

**Reply by smoothquark** [reply]
Running a search for Beyoncé on listen.tidal.com generates > 40 albums. I have noticed quite a few other inconsistencies too. For example, the artist "Sylvia" (Tidal id: 4607, MusicBrainz id: 47e5e2ec-7a3f-4dca-ab11-0e99362acbe4) shows albums from other Sylvias, but I presume this is due to errors within the metadata used by Tidal rather than the API.

**Reply by dylmye** [reply]
I think it's a paginated response with forced size, no? You use the `links.next` to fetch the next set of 20.

**Reply by aliaksei-taliuk** [reply]
Hi @smoothquark 

We're applying global cap of 20 items everywhere at this moment. But you definitely can iterate over all items.

For example, If you want iterate over an [artist albums](https://developer.tidal.com/apiref?spec=catalogue-v2&ref=get-artist-albums-v2) you must pick this particular relationship endpoint and use **links -> next** cursor you're getting in the response in order to request the next chunk of items.

**Reply by aliaksei-taliuk** [reply]
@dylmye is right, thanks for replying!

**Reply by smoothquark** [reply]
Ah, thank you! I've just seen the links->next after you pointed it out to me. I knew I must have been missing something. I don't think this is documented in the API documentation? I shall try it a bit later, and hopefully, everything will be OK. Also, I just realised that my query is off topic.

**Reply by aliaksei-taliuk** [reply]
>  I don't think this is documented in the API documentation?

It's part of the [JSON:API spec](https://jsonapi.org/format/).

**Reply by smoothquark** [reply]
Thank you. I was not aware that it was part of the JSON API spec.

**Comment by rotemdoron** [comment]
Joining all the requests, modify playlist or adding a favorite track to the use collection will be awesome!

**Comment by padraigf** [comment]
Hey guys,
I'm also interested in being able to write to playlists via the API.

My use-case:
I was trying to recreate a feature from Spotify, where you can hide a song on a public-playlist (or always skip it).
I find this feature very useful, and I notice it doesn't exist in Tidal, so I was trying to imitate it by copying the public-playlist to a private one, and then was hoping to manipulate it via the API (but I see I can't currently write to playlists).

Of course if Tidal implemented the 'hide' feature, that would save me trying to imitate it!

**Comment by tinacious** [comment]
With the risk of sounding like I'm parrotting everyone else in this thread, I too would find it very helpful to be able to manipulate playlists via the REST API.

My use case is that I am trying to migrate my Spotify playlists to Tidal. I'm not sure I want to even use the service yet but I'd like to have my playlists migrated over before I begin my trial. 

I do not want to use paid services like TuneMyMusic which have varying degrees of accuracy. Cleaning up the mess they make programmatically would also not be possible considering there are no delete endpoints. I'd like to build my own tooling for this to control the accuracy of the import.

Please enable the following actions in your REST API:

- Create playlists
- Add items to playlists
- Delete items from playlists
- Delete playlists
- Modify playlist privacy (e.g. public or private) either at the create stage or at the update stage would be fine

Those I think would be bare minimum support for playlists.

Attaching screenshots of Spotify and Apple Music documentation for reference.

Thanks

<img width="1650" alt="playlist apis - apple" src="https://github.com/user-attachments/assets/adca3f93-4af4-4fb4-9f09-d8c3c86a7140" />
<img width="1176" alt="image" src="https://github.com/user-attachments/assets/480b589c-3b42-4023-8b1a-b9c1b3805efe" />

**Reply by devsales** [reply]
"My use case is that I am trying to migrate my Spotify playlists to Tidal. I'm not sure I want to even use the service yet but I'd like to have my playlists migrated over before I begin my trial.

I do not want to use paid services like TuneMyMusic which have varying degrees of accuracy. Cleaning up the mess they make programmatically would also not be possible considering there are no delete endpoints. I'd like to build my own tooling for this to control the accuracy of the import."

Yes, this exactly! Thank you! 👍

**Comment by chad-williamson** [comment]
Bump. Just trying out switching from Spotify to Tidal, but TuneMyMusic did not maintain the order of my tracks.  Really need an API to manipulate tracks and playlists.

**Reply by januszmk** [reply]
@chad-williamson from what I saw, manipulating playlists is already possible https://tidal-music.github.io/tidal-api-reference/#/playlists/post_playlists__id__relationships_items. last time I tried using the api, it did throttle a lot (like only few requests per few seconds)

**Comment by SetlistGuy** [comment]
Hello, 
Definitely going to need to upload and verify  1x1 thumbnails to playlists. 
🤘🏼🤘🤘🏿🤘🏻🤘🏽🤘🏾
